bitcoin/src/amount.rs

```diff
@@ -1242,17 +1253,224 @@ pub mod serde {
+        }
+        fn visit_i16<E: serde::de::Error>(self, v: i16) -> Result<Self::Value, E> {
+            let pos = TryFrom::try_from(v).map_err(|_| E::custom(ParseAmountError::Negative))?;
+            Ok(Amount::from_sat(pos))
+        }
+        fn visit_i32<E: serde::de::Error>(self, v: i32) -> Result<Self::Value, E> {
+            let pos = TryFrom::try_from(v).map_err(|_| E::custom(ParseAmountError::Negative))?;
+            Ok(Amount::from_sat(pos))
+        }
+        fn visit_i64<E: serde::de::Error>(self, v: i64) -> Result<Self::Value, E> {
+            let pos = TryFrom::try_from(v).map_err(|_| E::custom(ParseAmountError::Negative))?;
+            Ok(Amount::from_sat(pos))
+        }
+        fn visit_u8<E: serde::de::Error>(self, v: u8) -> Result<Self::Value, E> {
+            Ok(Amount::from_sat(v as u64))
+        }
+        fn visit_u16<E: serde::de::Error>(self, v: u16) -> Result<Self::Value, E> {
+            Ok(Amount::from_sat(v as u64))
+        }
+        fn visit_u32<E: serde::de::Error>(self, v: u32) -> Result<Self::Value, E> {
+            Ok(Amount::from_sat(v as u64))
+        }
+        fn visit_u64<E: serde::de::Error>(self, v: u64) -> Result<Self::Value, E> {
+            Ok(Amount::from_sat(v))
+        }
+        fn visit_f32<E: serde::de::Error>(self, v: f32) -> Result<Self::Value, E> {
+            Amount::from_float_in(v as f64, Denomination::Satoshi).map_err(E::custom)
+        }
+        fn visit_f64<E: serde::de::Error>(self, v: f64) -> Result<Self::Value, E> {
+            Amount::from_float_in(v, Denomination::Satoshi).map_err(E::custom)
+        }
+    }
+
+    /// A serde Visitor for visiting signed amounts expressed in satoshi.
+    pub struct SignedSatVisitor;
+
+    impl<'de> serde::de::Visitor<'de> for SignedSatVisitor {
+        type Value = SignedAmount;
+        fn expecting(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            write!(f, "a signed Bitcoin amount in satoshi")
+        }
+        fn visit_i8<E: serde::de::Error>(self, v: i8) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
+        }
+        fn visit_i16<E: serde::de::Error>(self, v: i16) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
+        }
+        fn visit_i32<E: serde::de::Error>(self, v: i32) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
+        }
+        fn visit_i64<E: serde::de::Error>(self, v: i64) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v))
+        }
+        fn visit_u8<E: serde::de::Error>(self, v: u8) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
+        }
+        fn visit_u16<E: serde::de::Error>(self, v: u16) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
+        }
+        fn visit_u32<E: serde::de::Error>(self, v: u32) -> Result<Self::Value, E> {
+            Ok(SignedAmount::from_sat(v as i64))
```

Comment:
IDK how y'all feel about macros, but i have a strong suspicion that a macro would help reduce duplication of all these visitor methods for sized integers.

Reply:
We do use macros but we can't generate `visit_u32` from just `u32`, so it'd still need to be something like `ints!((u8, visit_u8), (u16, visit_u16))`... I wouldn't mind though.